allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py

```diff
@@ -0,0 +1,24 @@
+import torch
+
+from allennlp.common import Registrable
+
+
+class Seq2SeqDecoder(torch.nn.Module, Registrable):
+    # pylint: disable=abstract-method
+    """
+    A ``Seq2SeqDecoder`` is a ``Module`` that takes as input a result of ``Seq2SeqEncoder`` and returns a
+    new sequence of vectors.
+
+    Decoder should implement function to perform a single decoding step.
+    """
+
+    def __init__(self, output_dim):
```

Comment:
I don't think `output_dim` should be passed to `__init__`. Usually the decoder will output a `(batch_size, sequence_length)`, so `output_dim` would be `sequence_length`, but `sequence_length` is variable.

Reply:
Since we agreed, that decoder should be separated from Vocabulary and all Embeddings, we can't decode actual output token id inside decoder. We can only output predicted target embedding which should have dimensionality `output_dim` (or as I renamed it later `decoding_dim`). It's also important to return embedding for `BeamSearch`, because it may chose different indexes based on random search policy.

Reply:
IMHO a `Seq2SeqDecoder` should be in charge of outputting an entire sequence, not just a single hidden state or prediction from one timestep, otherwise it should probably be named differently. I'm also not convinced that this abstraction for a "seq2step" is general enough. As is, it wouldn't exactly work for CopyNet: https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/copynet_seq2seq.py

Reply:
Ok, good point on CopyNet. I will think about how to incorporate it into new decoder implementation. Agreed that current `Seq2SeqDecoder` should be called differently